operators/src/opencl/cl_program.rs

```diff
@@ -0,0 +1,1743 @@
+}
+
+/// Whether the kernel iterates over pixels or features
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum IterationType {
+    Raster,            // 2D Kernel, width x height
+    VectorFeatures,    // 1d kernel width = number of features
+    VectorCoordinates, // 1d kernel width = number of coordinates
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub struct RasterArgument {
+    pub data_type: RasterDataType,
+}
+
+impl RasterArgument {
+    pub fn new(data_type: RasterDataType) -> Self {
+        Self { data_type }
+    }
+}
+
+#[derive(PartialEq, Clone, Debug)]
+pub struct VectorArgument {
+    pub vector_type: VectorDataType,
+    // TODO: merge columns and types into one type
+    pub columns: Vec<String>,
+    pub column_types: Vec<FeatureDataType>,
+    pub include_geo: bool,
+    pub include_time: bool,
+}
+
+impl VectorArgument {
+    pub fn new(
+        vector_type: VectorDataType,
+        columns: Vec<String>,
+        column_types: Vec<FeatureDataType>,
+        include_geo: bool,
+        include_time: bool,
+    ) -> Self {
+        Self {
+            vector_type,
+            columns,
+            column_types,
+            include_geo,
+            include_time,
+        }
+    }
+}
+
+/// Specifies in and output types of CL program and compiles the source into a reusable `CompiledCLProgram`
+pub struct CLProgram {
+    input_rasters: Vec<RasterArgument>,
+    output_rasters: Vec<RasterArgument>,
+    input_features: Vec<VectorArgument>,
+    output_features: Vec<VectorArgument>,
+    iteration_type: IterationType,
+}
+
+impl CLProgram {
+    pub fn new(iteration_type: IterationType) -> Self {
+        Self {
+            input_rasters: vec![],
+            output_rasters: vec![],
+            input_features: vec![],
+            output_features: vec![],
+            iteration_type,
+        }
+    }
+
+    pub fn add_input_raster(&mut self, raster: RasterArgument) {
+        self.input_rasters.push(raster);
+    }
+
+    pub fn add_output_raster(&mut self, raster: RasterArgument) {
+        self.output_rasters.push(raster);
+    }
+
+    pub fn add_input_features(&mut self, vector_type: VectorArgument) {
+        self.input_features.push(vector_type);
+    }
+
+    pub fn add_output_features(&mut self, vector_type: VectorArgument) {
+        self.output_features.push(vector_type);
+    }
+
+    fn create_type_definitions(&self) -> String {
+        let mut s = String::new();
+
+        if self.input_rasters.len() + self.output_rasters.len() == 0 {
+            return s;
+        }
+
+        s.push_str(
+            r####"
+typedef struct {
+	uint size[3];
+	double origin[3];
+	double scale[3];
+	double min, max, no_data;
+	ushort crs_code;
+	ushort has_no_data;
+} RasterInfo;
+
+#define R(t,x,y) t ## _data[y * t ## _info->size[0] + x]
+"####,
+        );
+
+        for (idx, raster) in self.input_rasters.iter().enumerate() {
+            s += &format!("typedef {} IN_TYPE{};\n", raster.data_type.ocl_type(), idx);
```

Comment:
can we move this to a header file and include it in the kernel?

Reply:
this is dynamically generated so I can't read it from a file. Also: I already include the source codes for the kernel and the typedes separately:
```
let program = ProgramBuilder::new()
            .src(typedefs)
            .src(source)
            .build(&ctx)?;
```